File: packages/capacitor-plugin/ios/Sources/LocalLLMPlugin/LocalLLM.swift
import Foundation
import FoundationModels

public enum LLMAvailability: String {
    case available = "available"
    case unsupported = "unsupported"
    case notEnabled = "notEnabled"
    case notReady = "notReady"
    case unavailable = "unavailable"
}

public struct LLMOptions {
    let temperature: Double?
    let maximiumOutputTokens: Int?
}

public struct LLMPromptOptions {
    let sessionId: String?
    let instructions: String?
    let options: LLMOptions?
    let prompt: String
}

public enum LocalLLMError: Error {
    case responseInProgress
    case sessionNotFound
}

public class LocalLLM {
    private var sessions: [String: LanguageModelSession] = [:]
    
    static func availability() -> LLMAvailability {
        let status = SystemLanguageModel.default.availability
        
        switch status {
        case .available:
            return .available
        case .unavailable(.deviceNotEligible):
            return .unsupported
        case .unavailable(.appleIntelligenceNotEnabled):
            return .notEnabled
        case .unavailable(.modelNotReady):
            return .notReady
        case .unavailable(_):
            return .unavailable
        }
    }
    
    func prompt(options: LLMPromptOptions) async throws -> String {
        var session: LanguageModelSession?
        
        if let sessionId = options.sessionId {
            session = sessions[sessionId]
        } else {
            session = LanguageModelSession(instructions: options.instructions)
        }
        
        guard let session = session else {
            throw LocalLLMError.sessionNotFound
        }
        
        if session.isResponding {
            throw LocalLLMError.responseInProgress
        }
        
        let response = try await session.respond(to: options.prompt)
        
        return response.content
    }
    
    func endSession(_ sessionId: String) {
        sessions[sessionId] = nil
    }
}

File: android/src/main/java/io/ionic/localllm/plugin/LocalLLM.kt
package io.ionic.localllm.plugin

import android.os.Build
import com.getcapacitor.Logger
import com.google.mlkit.genai.common.DownloadStatus
import com.google.mlkit.genai.common.FeatureStatus
import com.google.mlkit.genai.prompt.Generation

import com.google.mlkit.genai.prompt.GenerativeModel
import com.google.mlkit.genai.prompt.TextPart
import com.google.mlkit.genai.prompt.generateContentRequest

data class ChatSession(
    val instructions: String?,
    val history: MutableList<Pair<String, String>> = mutableListOf()
)

class LocalLLM(private val context: android.content.Context) {
    val model: GenerativeModel = Generation.getClient()
    private val sessions: MutableMap<String, ChatSession> = mutableMapOf()

    suspend fun availability(): LLMAvailability {
        val status = model.checkStatus()
        when (status) {
            FeatureStatus.UNAVAILABLE -> {
                return LLMAvailability.Unavailable
            }
            FeatureStatus.DOWNLOADABLE ->{
                return LLMAvailability.Downloadable
            }
            FeatureStatus.DOWNLOADING -> {
                return LLMAvailability.NotReady
            }
            FeatureStatus.AVAILABLE -> {
                return LLMAvailability.Available
            }
        }

        return LLMAvailability.Unavailable
    }

    suspend fun download() {
        model.download().collect { status ->
            when(status) {
                is DownloadStatus.DownloadStarted ->
                    Logger.debug("LocalLLM", "Gemini Nano download started")
                is DownloadStatus.DownloadProgress ->
                    Logger.debug("LocalLLM", "Gemini Nano ${status.totalBytesDownloaded} bytes downloaded")
                is DownloadStatus.DownloadCompleted ->
                    Logger.debug("LocalLLM", "Gemini Nano download completed")
                is DownloadStatus.DownloadFailed ->
                    Logger.debug("LocalLLM", "Gemini Nano download failed: ${status.e.message}")
            }
        }
    }

    suspend fun prompt(options: LLMPromptOptions): String {
        val sessionId = options.sessionId

        val fullPrompt = if (sessionId != null) {
            val session = sessions.getOrPut(sessionId) {
                ChatSession(instructions = options.instructions)
            }

            buildString {
                if (session.instructions != null) {
                    appendLine(session.instructions)
                    appendLine()
                }

                for ((userMsg, assistantMsg) in session.history) {
                    appendLine("User: $userMsg")
                    appendLine("Assistant: $assistantMsg")
                    appendLine()
                }

                append("User: ${options.prompt}")
            }
        } else {
            if (options.instructions != null) {
                "${options.instructions}\n\n${options.prompt}"
            } else {
                options.prompt
            }
        }

        val response = model.generateContent(
            generateContentRequest(
                TextPart(fullPrompt)
            ) {
                temperature = options.options?.temperature
                topK = 16
                maxOutputTokens = options.options?.maxOutputTokens
            }
        )

        val responseText = response.candidates.first().text

        if (sessionId != null) {
            sessions[sessionId]?.history?.add(Pair(options.prompt, responseText))
        }

        return responseText
    }

    fun endSession(sessionId: String) {
        sessions.remove(sessionId)
    }
}
